File: .skills/image-resize-helper.js
/**
 * Image Resize Helper - Project Agnostic Utility
 *
 * Standalone utility for resizing and processing images during asset generation.
 * Can read configuration from .asset-gen-config.json or work independently.
 *
 * Usage:
 *   node image-resize-helper.js <input-path> <output-path> <width> <height> [strategy]
 *
 * Strategies:
 *   - contain-centered: Resize to fit within dimensions, center with padding (for icons)
 *   - cover-crop: Resize to cover dimensions, crop excess (for textures/backgrounds)
 *   - stretch: Resize to exact dimensions (may distort)
 *
 * Dependencies:
 *   - sharp (npm install sharp)
 */

const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

// Strategy constants
const STRATEGIES = {
  CONTAIN_CENTERED: 'contain-centered',
  COVER_CROP: 'cover-crop',
  STRETCH: 'stretch'
};

/**
 * Resize an image using the specified strategy
 * @param {string} inputPath - Path to input image
 * @param {string} outputPath - Path to save output image
 * @param {number} targetWidth - Target width in pixels
 * @param {number} targetHeight - Target height in pixels
 * @param {string} strategy - Resize strategy (contain-centered, cover-crop, stretch)
 * @returns {Promise<object>} - Result object with metadata
 */
async function resizeImage(inputPath, outputPath, targetWidth, targetHeight, strategy = STRATEGIES.CONTAIN_CENTERED) {
  const input = sharp(inputPath);
  const metadata = await input.metadata();

  let pipeline;

  switch (strategy) {
    case STRATEGIES.CONTAIN_CENTERED:
      // Resize to fit within bounds, center with transparent padding
      pipeline = input
        .resize(targetWidth, targetHeight, {
          fit: 'contain',
          background: { r: 0, g: 0, b: 0, alpha: 0 }
        });
      break;

    case STRATEGIES.COVER_CROP:
      // Resize to cover bounds, crop excess from center
      pipeline = input
        .resize(targetWidth, targetHeight, {
          fit: 'cover',
          position: 'center'
        });
      break;

    case STRATEGIES.STRETCH:
      // Resize to exact dimensions (may distort aspect ratio)
      pipeline = input
        .resize(targetWidth, targetHeight, {
          fit: 'fill'
        });
      break;

    default:
      throw new Error(`Unknown resize strategy: ${strategy}`);
  }

  // Determine output format from file extension
  const ext = path.extname(outputPath).toLowerCase();

  if (ext === '.jpg' || ext === '.jpeg') {
    pipeline = pipeline.jpeg({ quality: 85 });
  } else if (ext === '.png') {
    pipeline = pipeline.png({ compressionLevel: 9 });
  } else if (ext === '.webp') {
    pipeline = pipeline.webp({ quality: 85 });
  }

  await pipeline.toFile(outputPath);

  const outputMetadata = await sharp(outputPath).metadata();

  return {
    success: true,
    input: {
      path: inputPath,
      width: metadata.width,
      height: metadata.height,
      format: metadata.format
    },
    output: {
      path: outputPath,
      width: outputMetadata.width,
      height: outputMetadata.height,
      format: outputMetadata.format,
      size: fs.statSync(outputPath).size
    },
    strategy
  };
}

/**
 * Center an icon within a canvas
 * @param {string} inputPath - Path to input image
 * @param {string} outputPath - Path to save output image
 * @param {number} canvasSize - Size of the square canvas
 * @param {number} padding - Padding percentage (0-1)
 * @returns {Promise<object>} - Result object with metadata
 */
async function centerIcon(inputPath, outputPath, canvasSize, padding = 0.1) {
  const input = sharp(inputPath);
  const metadata = await input.metadata();

  // Calculate the size to fit the icon with padding
  const maxIconSize = Math.floor(canvasSize * (1 - padding * 2));

  // Resize the icon to fit within max size while maintaining aspect ratio
  const resizedBuffer = await input
    .resize(maxIconSize, maxIconSize, {
      fit: 'inside',
      withoutEnlargement: false
    })
    .toBuffer();

  // Get the resized dimensions
  const resizedMeta = await sharp(resizedBuffer).metadata();

  // Calculate padding to center
  const left = Math.floor((canvasSize - resizedMeta.width) / 2);
  const top = Math.floor((canvasSize - resizedMeta.height) / 2);

  // Create transparent canvas and composite the icon
  await sharp({
    create: {
      width: canvasSize,
      height: canvasSize,
      channels: 4,
      background: { r: 0, g: 0, b: 0, alpha: 0 }
    }
  })
    .composite([{
      input: resizedBuffer,
      left,
      top
    }])
    .png({ compressionLevel: 9 })
    .toFile(outputPath);

  return {
    success: true,
    input: {
      path: inputPath,
      width: metadata.width,
      height: metadata.height
    },
    output: {
      path: outputPath,
      width: canvasSize,
      height: canvasSize,
      size: fs.statSync(outputPath).size
    },
    strategy: 'center-icon',
    padding
  };
}

/**
 * Remove white/near-white colors from an image (make them transparent)
 * @param {string} inputPath - Path to input image
 * @param {string} outputPath - Path to save output image
 * @param {number} threshold - White threshold (0-255), pixels above this become transparent
 * @returns {Promise<object>} - Result object with metadata
 */
async function removeWhite(inputPath, outputPath, threshold = 240) {
  const input = sharp(inputPath);
  const metadata = await input.metadata();

  // Get raw pixel data
  const { data, info } = await input
    .raw()
    .ensureAlpha()
    .toBuffer({ resolveWithObject: true });

  // Process pixels
  for (let i = 0; i < data.length; i += 4) {
    const r = data[i];
    const g = data[i + 1];
    const b = data[i + 2];

    // If pixel is near-white, make it transparent
    if (r >= threshold && g >= threshold && b >= threshold) {
      data[i + 3] = 0; // Set alpha to 0
    }
  }

  await sharp(data, {
    raw: {
      width: info.width,
      height: info.height,
      channels: 4
    }
  })
    .png({ compressionLevel: 9 })
    .toFile(outputPath);

  return {
    success: true,
    input: { path: inputPath },
    output: {
      path: outputPath,
      size: fs.statSync(outputPath).size
    },
    threshold
  };
}

/**
 * Erode alpha channel to remove white halos
 * @param {string} inputPath - Path to input image
 * @param {string} outputPath - Path to save output image
 * @param {number} pixels - Number of pixels to erode
 * @returns {Promise<object>} - Result object with metadata
 */
async function erodeAlpha(inputPath, outputPath, pixels = 1) {
  const input = sharp(inputPath);

  // Use a simple approach: resize down then up to erode edges
  const metadata = await input.metadata();

  const erosionFactor = 1 - (pixels * 2 / Math.min(metadata.width, metadata.height));
  const shrunkWidth = Math.floor(metadata.width * erosionFactor);
  const shrunkHeight = Math.floor(metadata.height * erosionFactor);

  await input
    .resize(shrunkWidth, shrunkHeight, { fit: 'inside' })
    .resize(metadata.width, metadata.height, {
      fit: 'contain',
      background: { r: 0, g: 0, b: 0, alpha: 0 }
    })
    .png({ compressionLevel: 9 })
    .toFile(outputPath);

  return {
    success: true,
    input: { path: inputPath },
    output: {
      path: outputPath,
      size: fs.statSync(outputPath).size
    },
    erosionPixels: pixels
  };
}

/**
 * Batch resize all assets in a folder according to config
 * @param {string} configPath - Path to .asset-gen-config.json
 * @param {string} inputFolder - Folder containing assets to resize
 * @param {string} outputFolder - Folder to save resized assets (can be same as input)
 * @returns {Promise<object>} - Results for each asset
 */
async function batchResize(configPath, inputFolder, outputFolder) {
  const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
  const results = {};

  for (const [assetType, assetConfig] of Object.entries(config.assetTypes)) {
    if (assetType.startsWith('_')) continue; // Skip comments

    const inputPath = path.join(inputFolder, assetConfig.filename);
    const outputPath = path.join(outputFolder, assetConfig.filename);

    if (!fs.existsSync(inputPath)) {
      results[assetType] = { success: false, error: 'File not found', path: inputPath };
      continue;
    }

    // Determine strategy based on asset type
    let strategy = STRATEGIES.CONTAIN_CENTERED;
    if (assetConfig.filename.includes('texture') || assetConfig.filename.includes('background')) {
      strategy = STRATEGIES.COVER_CROP;
    }

    // Check postProcessing for explicit strategy
    if (assetConfig.postProcessing) {
      for (const step of assetConfig.postProcessing) {
        if (step.includes('contain-centered')) {
          strategy = STRATEGIES.CONTAIN_CENTERED;
        } else if (step.includes('cover-crop')) {
          strategy = STRATEGIES.COVER_CROP;
        }
      }
    }

    try {
      const result = await resizeImage(
        inputPath,
        outputPath,
        assetConfig.dimensions.width,
        assetConfig.dimensions.height,
        strategy
      );
      results[assetType] = result;
    } catch (error) {
      results[assetType] = { success: false, error: error.message, path: inputPath };
    }
  }

  return results;
}

// CLI interface
if (require.main === module) {
  const args = process.argv.slice(2);

  if (args.length < 4 && args[0] !== '--batch' && args[0] !== '--remove-white' && args[0] !== '--center' && args[0] !== '--erode') {
    console.log(`
Image Resize Helper - Project Agnostic Utility

Usage: node image-resize-helper.js <input-path> <output-path> <width> <height> [strategy]

Strategies:
  contain-centered  - Resize to fit, center with padding (default for icons)
  cover-crop        - Resize to cover, crop excess (default for textures)
  stretch           - Resize to exact dimensions

Examples:
  node image-resize-helper.js input.png output.png 256 256 contain-centered
  node image-resize-helper.js texture.jpg texture.jpg 512 512 cover-crop

Special Commands:
  node image-resize-helper.js --batch <config-path> <input-folder> <output-folder>
  node image-resize-helper.js --remove-white <input> <output> [threshold=240]
  node image-resize-helper.js --center <input> <output> <size> [padding=0.1]
  node image-resize-helper.js --erode <input> <output> [pixels=1]
`);
    process.exit(1);
  }

  (async () => {
    try {
      if (args[0] === '--batch') {
        const results = await batchResize(args[1], args[2], args[3]);
        console.log(JSON.stringify(results, null, 2));
      } else if (args[0] === '--remove-white') {
        const result = await removeWhite(args[1], args[2], parseInt(args[3]) || 240);
        console.log(JSON.stringify(result, null, 2));
      } else if (args[0] === '--center') {
        const result = await centerIcon(args[1], args[2], parseInt(args[3]), parseFloat(args[4]) || 0.1);
        console.log(JSON.stringify(result, null, 2));
      } else if (args[0] === '--erode') {
        const result = await erodeAlpha(args[1], args[2], parseInt(args[3]) || 1);
        console.log(JSON.stringify(result, null, 2));
      } else {
        const [inputPath, outputPath, width, height, strategy] = args;
        const result = await resizeImage(inputPath, outputPath, parseInt(width), parseInt(height), strategy || STRATEGIES.CONTAIN_CENTERED);
        console.log(JSON.stringify(result, null, 2));
      }
    } catch (error) {
      console.error('Error:', error.message);
      process.exit(1);
    }
  })();
}

module.exports = {
  resizeImage,
  centerIcon,
  removeWhite,
  erodeAlpha,
  batchResize,
  STRATEGIES
};

File: .skills/tech-artist-generate-asset/image-resize-helper.py
#!/usr/bin/env python3
"""
Image Resize Helper - Project Agnostic Utility

Standalone utility for resizing and processing images during asset generation.
Can read configuration from .asset-gen-config.json or work independently.

Usage:
    python image-resize-helper.py <input-path> <output-path> <width> <height> [strategy]

Strategies:
    - contain-centered: Resize to fit within dimensions, center with padding (for icons)
    - cover-crop: Resize to cover dimensions, crop excess (for textures/backgrounds)
    - stretch: Resize to exact dimensions (may distort)

Dependencies:
    pip install Pillow
"""

import sys
import os
import json
from pathlib import Path
from PIL import Image, ImageOps


# Strategy constants
STRATEGIES = {
    'CONTAIN_CENTERED': 'contain-centered',
    'COVER_CROP': 'cover-crop',
    'STRETCH': 'stretch'
}


def resize_image(input_path, output_path, target_width, target_height, strategy='contain-centered'):
    """
    Resize an image using the specified strategy

    Args:
        input_path: Path to input image
        output_path: Path to save output image
        target_width: Target width in pixels
        target_height: Target height in pixels
        strategy: Resize strategy (contain-centered, cover-crop, stretch)

    Returns:
        dict: Result object with metadata
    """
    image = Image.open(input_path)
    original_size = image.size
    original_format = image.format

    if strategy == STRATEGIES['CONTAIN_CENTERED']:
        # Resize to fit within bounds, center with transparent padding
        image.thumbnail((target_width, target_height), Image.Resampling.LANCZOS)

        # Create transparent canvas
        if image.mode == 'RGBA':
            canvas = Image.new('RGBA', (target_width, target_height), (0, 0, 0, 0))
        else:
            canvas = Image.new('RGB', (target_width, target_height), (255, 255, 255))

        # Center image on canvas
        offset = ((target_width - image.width) // 2, (target_height - image.height) // 2)
        canvas.paste(image, offset)
        image = canvas

    elif strategy == STRATEGIES['COVER_CROP']:
        # Resize to cover bounds, crop excess from center
        image = ImageOps.fit(image, (target_width, target_height), Image.Resampling.LANCZOS, centering=(0.5, 0.5))

    elif strategy == STRATEGIES['STRETCH']:
        # Resize to exact dimensions (may distort aspect ratio)
        image = image.resize((target_width, target_height), Image.Resampling.LANCZOS)

    else:
        raise ValueError(f"Unknown resize strategy: {strategy}")

    # Determine output format from file extension
    ext = Path(output_path).suffix.lower()

    if ext in ['.jpg', '.jpeg']:
        if image.mode == 'RGBA':
            # Convert RGBA to RGB for JPEG
            rgb_image = Image.new('RGB', image.size, (255, 255, 255))
            rgb_image.paste(image, mask=image.split()[3] if image.mode == 'RGBA' else None)
            image = rgb_image
        image.save(output_path, 'JPEG', quality=85, optimize=True)
    elif ext == '.png':
        image.save(output_path, 'PNG', optimize=True)
    elif ext == '.webp':
        image.save(output_path, 'WEBP', quality=85)
    else:
        image.save(output_path)

    output_size = os.path.getsize(output_path)

    return {
        'success': True,
        'input': {
            'path': input_path,
            'width': original_size[0],
            'height': original_size[1],
            'format': original_format
        },
        'output': {
            'path': output_path,
            'width': image.width,
            'height': image.height,
            'size': output_size
        },
        'strategy': strategy
    }


def center_icon(input_path, output_path, canvas_size, padding=0.1):
    """
    Center an icon within a canvas

    Args:
        input_path: Path to input image
        output_path: Path to save output image
        canvas_size: Size of the square canvas
        padding: Padding percentage (0-1)

    Returns:
        dict: Result object with metadata
    """
    image = Image.open(input_path)
    original_size = image.size

    # Calculate the size to fit the icon with padding
    max_icon_size = int(canvas_size * (1 - padding * 2))

    # Resize the icon to fit within max size while maintaining aspect ratio
    image.thumbnail((max_icon_size, max_icon_size), Image.Resampling.LANCZOS)

    # Create transparent canvas
    canvas = Image.new('RGBA', (canvas_size, canvas_size), (0, 0, 0, 0))

    # Calculate padding to center
    left = (canvas_size - image.width) // 2
    top = (canvas_size - image.height) // 2

    # Paste icon on canvas
    canvas.paste(image, (left, top), image if image.mode == 'RGBA' else None)

    # Save as PNG
    canvas.save(output_path, 'PNG', optimize=True)

    output_size = os.path.getsize(output_path)

    return {
        'success': True,
        'input': {
            'path': input_path,
            'width': original_size[0],
            'height': original_size[1]
        },
        'output': {
            'path': output_path,
            'width': canvas_size,
            'height': canvas_size,
            'size': output_size
        },
        'strategy': 'center-icon',
        'padding': padding
    }


def remove_white(input_path, output_path, threshold=240):
    """
    Remove white/near-white colors from an image (make them transparent)

    Args:
        input_path: Path to input image
        output_path: Path to save output image
        threshold: White threshold (0-255), pixels above this become transparent

    Returns:
        dict: Result object with metadata
    """
    image = Image.open(input_path).convert('RGBA')
    pixels = image.load()

    for y in range(image.height):
        for x in range(image.width):
            r, g, b, a = pixels[x, y]
            if r >= threshold and g >= threshold and b >= threshold:
                pixels[x, y] = (r, g, b, 0)

    image.save(output_path, 'PNG', optimize=True)

    output_size = os.path.getsize(output_path)

    return {
        'success': True,
        'input': {'path': input_path},
        'output': {
            'path': output_path,
            'size': output_size
        },
        'threshold': threshold
    }


def erode_alpha(input_path, output_path, pixels=1):
    """
    Erode alpha channel to remove white halos

    Args:
        input_path: Path to input image
        output_path: Path to save output image
        pixels: Number of pixels to erode

    Returns:
        dict: Result object with metadata
    """
    image = Image.open(input_path).convert('RGBA')
    original_size = image.size

    # Simple erosion: resize down then up
    erosion_factor = 1 - (pixels * 2 / min(image.width, image.height))
    shrunk_width = int(image.width * erosion_factor)
    shrunk_height = int(image.height * erosion_factor)

    # Resize down
    shrunk = image.resize((shrunk_width, shrunk_height), Image.Resampling.LANCZOS)

    # Create transparent canvas of original size
    canvas = Image.new('RGBA', original_size, (0, 0, 0, 0))

    # Center the shrunk image
    offset = ((original_size[0] - shrunk_width) // 2, (original_size[1] - shrunk_height) // 2)
    canvas.paste(shrunk, offset, shrunk)

    canvas.save(output_path, 'PNG', optimize=True)

    output_size = os.path.getsize(output_path)

    return {
        'success': True,
        'input': {'path': input_path},
        'output': {
            'path': output_path,
            'size': output_size
        },
        'erosionPixels': pixels
    }


def threshold_to_black(input_path, output_path, alpha_threshold=10):
    """
    Convert all non-transparent pixels to pure black (#000000) for clean engine tinting.
    This is the "Sticker & Contrast" strategy for icon generation.

    Use this after background removal to create perfect black masks that game engines
    can tint with any color without artifacts from "almost black" or colored pixels.

    Args:
        input_path: Path to input image (must have alpha channel)
        output_path: Path to save output image
        alpha_threshold: Alpha value below which pixels are considered transparent (0-255)

    Returns:
        dict: Result object with metadata
    """
    image = Image.open(input_path).convert('RGBA')
    pixels = image.load()

    pixels_converted = 0

    for y in range(image.height):
        for x in range(image.width):
            r, g, b, a = pixels[x, y]

            # If pixel is not transparent (alpha above threshold)
            if a > alpha_threshold:
                # Convert to pure black, preserve alpha
                pixels[x, y] = (0, 0, 0, a)
                pixels_converted += 1

    image.save(output_path, 'PNG', optimize=True)

    output_size = os.path.getsize(output_path)

    return {
        'success': True,
        'input': {'path': input_path},
        'output': {
            'path': output_path,
            'size': output_size
        },
        'alphaThreshold': alpha_threshold,
        'pixelsConverted': pixels_converted,
        'rationale': 'Pure black mask for clean game engine tinting'
    }


def batch_resize(config_path, input_folder, output_folder):
    """
    Batch resize all assets in a folder according to config

    Args:
        config_path: Path to .asset-gen-config.json
        input_folder: Folder containing assets to resize
        output_folder: Folder to save resized assets (can be same as input)

    Returns:
        dict: Results for each asset
    """
    with open(config_path, 'r') as f:
        config = json.load(f)

    results = {}

    for asset_type, asset_config in config['assetTypes'].items():
        if asset_type.startswith('_'):
            continue

        input_path = os.path.join(input_folder, asset_config['filename'])
        output_path = os.path.join(output_folder, asset_config['filename'])

        if not os.path.exists(input_path):
            results[asset_type] = {'success': False, 'error': 'File not found', 'path': input_path}
            continue

        # Determine strategy based on asset type
        strategy = STRATEGIES['CONTAIN_CENTERED']
        if 'texture' in asset_config['filename'] or 'background' in asset_config['filename']:
            strategy = STRATEGIES['COVER_CROP']

        # Check postProcessing for explicit strategy
        if 'postProcessing' in asset_config:
            for step in asset_config['postProcessing']:
                if 'contain-centered' in step:
                    strategy = STRATEGIES['CONTAIN_CENTERED']
                elif 'cover-crop' in step:
                    strategy = STRATEGIES['COVER_CROP']

        try:
            result = resize_image(
                input_path,
                output_path,
                asset_config['dimensions']['width'],
                asset_config['dimensions']['height'],
                strategy
            )
            results[asset_type] = result
        except Exception as error:
            results[asset_type] = {'success': False, 'error': str(error), 'path': input_path}

    return results


def main():
    """Main CLI interface"""
    args = sys.argv[1:]

    if len(args) < 4 and args[0] not in ['--batch', '--remove-white', '--center', '--erode', '--threshold-black']:
        print("""
Image Resize Helper - Project Agnostic Utility

Usage: python image-resize-helper.py <input-path> <output-path> <width> <height> [strategy]

Strategies:
  contain-centered  - Resize to fit, center with padding (default for icons)
  cover-crop        - Resize to cover, crop excess (default for textures)
  stretch           - Resize to exact dimensions

Examples:
  python image-resize-helper.py input.png output.png 256 256 contain-centered
  python image-resize-helper.py texture.jpg texture.jpg 512 512 cover-crop

Special Commands:
  python image-resize-helper.py --batch <config-path> <input-folder> <output-folder>
  python image-resize-helper.py --remove-white <input> <output> [threshold=240]
  python image-resize-helper.py --center <input> <output> <size> [padding=0.1]
  python image-resize-helper.py --erode <input> <output> [pixels=1]
  python image-resize-helper.py --threshold-black <input> <output> [alpha-threshold=10]
""")
        sys.exit(1)

    try:
        if args[0] == '--batch':
            results = batch_resize(args[1], args[2], args[3])
            print(json.dumps(results, indent=2))
        elif args[0] == '--remove-white':
            result = remove_white(args[1], args[2], int(args[3]) if len(args) > 3 else 240)
            print(json.dumps(result, indent=2))
        elif args[0] == '--center':
            result = center_icon(args[1], args[2], int(args[3]), float(args[4]) if len(args) > 4 else 0.1)
            print(json.dumps(result, indent=2))
        elif args[0] == '--erode':
            result = erode_alpha(args[1], args[2], int(args[3]) if len(args) > 3 else 1)
            print(json.dumps(result, indent=2))
        elif args[0] == '--threshold-black':
            result = threshold_to_black(args[1], args[2], int(args[3]) if len(args) > 3 else 10)
            print(json.dumps(result, indent=2))
        else:
            input_path, output_path, width, height = args[:4]
            strategy = args[4] if len(args) > 4 else 'contain-centered'
            result = resize_image(input_path, output_path, int(width), int(height), strategy)
            print(json.dumps(result, indent=2))
    except Exception as error:
        print(f'Error: {error}')
        sys.exit(1)


if __name__ == '__main__':
    main()
